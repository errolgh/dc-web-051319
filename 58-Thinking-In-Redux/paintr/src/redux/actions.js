//write functions and create and return action objects for us

function changeSearch(event){
  return {type:"CHANGE_SEARCH", payload: event.target.value}
}

function voteForPainting(paintingId){
  return {type:"VOTE_FOR_PAINTING", payload: paintingId}
}

export {changeSearch, voteForPainting}
