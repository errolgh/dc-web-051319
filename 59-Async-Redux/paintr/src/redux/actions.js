// action creators
const URL = 'http://localhost:3000/paintings/'

function changeSearchText(value) {
  return { type: "CHANGE_SEARCH_TEXT", value: value };
}

function voted(paintingId) {
  return { type: "INCREASE_VOTES", paintingId };
}

function updatePainting({ title, name, birthday, deathday, paintingId }) {
  return {
    type: "UPDATE_PAINTING",
    payload: { title, name, birthday, deathday },
    paintingId
  };
}

function fetchedPainting(paintings){
  return {type:"FETCHED_PAINTINGS", payload: paintings}
}

function isLoading(){
  return {type: "LOADING_PAINTINGS"}
}

function fetchingPainting(){
  return (dispatch) => {
    dispatch(isLoading())
    fetch(URL)
    .then(res => res.json())
    .then(paintings => {
      dispatch( fetchedPainting(paintings) )
    })
  }
}
//Problem 1: we need to wait for the fetch to happen before returning action
//Problem 2: we don't have access to dispatch

function voting(paintingId){
  return (dispatch, getState) => {
    let oldVotes = getState().paintings.find(p => p.id === paintingId).votes
    fetch(`${URL}/${paintingId}`, {
      method: "PATCH",
      headers: {
        "Content-Type" : "application/json",
        "Accept" : "application/json"
      },
      body: JSON.stringify({
        votes : oldVotes + 1
      })
    }).then(res => {
      dispatch(voted(paintingId))
    })
  }
}

export { changeSearchText, voted, updatePainting, fetchedPainting,
  fetchingPainting, voting };
