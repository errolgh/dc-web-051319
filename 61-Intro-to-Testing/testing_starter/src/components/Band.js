import React, { Component } from 'react';
import { connect } from 'react-redux';

class Band extends Component {
    render() {
        return (
            <div className='band-card'>
                <h2>{this.props.band.name}</h2>
                <img src={this.props.band.img_url} />
                <div>Genre: {this.props.band.genre}</div>
                <button onClick={() => this.props.deleteBand(this.props.band.id)} >Delete</button>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        deleteBand: (bandId) => dispatch({ type: 'DELETE_BAND', payload: bandId})
    }
}

export default connect(null, mapDispatchToProps)(Band)