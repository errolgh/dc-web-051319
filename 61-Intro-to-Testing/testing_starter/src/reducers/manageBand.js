let idTracker = 1;
let randomTracker = 1;

export default function manageBand(state = { bands: [] }, action){
    switch (action.type) {
        case 'ADD_BAND':
            let band = { id: idTracker, img_url: `https://loremflickr.com/320/240?random=${randomTracker}''`, name: action.payload.name, genre: action.payload.genre }
            idTracker++
            randomTracker++
            console.log(band)
            return {...state, bands: [...state.bands, band]}
        case 'DELETE_BAND':
            let filteredBands = state.bands.filter(band => band.id !== action.payload)
            return {...state, bands: filteredBands }
        default:
            return state
    }
}