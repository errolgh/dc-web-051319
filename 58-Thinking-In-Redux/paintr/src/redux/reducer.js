import paintingsData from "../paintings.json"
import {combineReducers} from 'redux'

let initialState = {
  paintings: paintingsData.paintings,
  searchText: ""
}

const searchTextReducer = (oldState=initialState.searchText, action) => {
  switch (action.type) {
    case "CHANGE_SEARCH":
      return action.payload
    default:
      return oldState
  }
}

const paintingsReducer = (oldState=initialState.paintings, action) => {
  switch (action.type) {
    case "VOTE_FOR_PAINTING":
      //create a new paintings, same length as the old painting,
      //but with the one painting given another vote
      let newPaintings = oldState.map(p => {
        if(p.id === action.payload){
          return {...p, votes: p.votes + 1}
        }else{
          return p
        }
      })
      return newPaintings
    default:
      return oldState
  }
}

//maps the state : reducer
let rootReducer = combineReducers({
  paintings: paintingsReducer,
  searchText: searchTextReducer
})

export default rootReducer
