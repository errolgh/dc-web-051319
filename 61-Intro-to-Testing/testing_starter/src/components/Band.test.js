import React from 'react'
import { createStore } from 'redux';
import manageBand from '../reducers/manageBand';
import Band from './Band';
import renderer from 'react-test-renderer';
import { jsxElement, exportAllDeclaration } from '@babel/types';

let mockStore = createStore(manageBand)

let mockProps = {
    band: {
        id: 1,
        img_url: null,
        name: 'The Smiths',
        genre: 'brit pop',
        props: jest.fn()
    }
}

describe ('Band', ()=>{
    it ('renders correctly', () => { 
        const tree = renderer
        .create(<Band {...mockProps} store={mockStore} />)
        .toJSON();
        expect(tree).toMatchSnapshot();
    })
})



