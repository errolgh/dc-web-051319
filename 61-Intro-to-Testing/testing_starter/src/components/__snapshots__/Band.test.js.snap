// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Band renders correctly 1`] = `
<div
  className="band-card"
>
  <h2>
    The Smiths
  </h2>
  <img
    src={null}
  />
  <div>
    Genre: 
    brit pop
  </div>
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
